File: articles/design-thinking-ideation/go/idea_value_engine.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

type Idea struct {
	Name          string
	Desirability  float64
	Feasibility   float64
	Novelty       float64
	EquityValue   float64
	LearningValue float64
	ResidualRisk  float64
	Value         float64
}

func parseFloat(value string) float64 {
	parsed, err := strconv.ParseFloat(value, 64)
	if err != nil {
		log.Fatalf("unable to parse numeric value %q: %v", value, err)
	}
	return parsed
}

func ideaValue(idea Idea) float64 {
	return 0.24*idea.Desirability +
		0.18*idea.Feasibility +
		0.18*idea.Novelty +
		0.18*idea.EquityValue +
		0.12*idea.LearningValue -
		0.10*idea.ResidualRisk
}

func loadIdeas(path string) []Idea {
	file, err := os.Open(path)
	if err != nil {
		log.Fatalf("unable to open input file: %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	rows, err := reader.ReadAll()
	if err != nil {
		log.Fatalf("unable to read CSV: %v", err)
	}

	var ideas []Idea

	for i, row := range rows {
		if i == 0 {
			continue
		}
		if len(row) < 8 {
			log.Fatalf("malformed row: %v", row)
		}

		idea := Idea{
			Name:          row[0],
			Desirability:  parseFloat(row[2]),
			Feasibility:   parseFloat(row[3]),
			Novelty:       parseFloat(row[4]),
			EquityValue:   parseFloat(row[5]),
			LearningValue: parseFloat(row[6]),
			ResidualRisk:  parseFloat(row[7]),
		}
		idea.Value = ideaValue(idea)
		ideas = append(ideas, idea)
	}

	return ideas
}

func main() {
	path := "../data/raw/idea_portfolio_raw.csv"
	if len(os.Args) > 1 {
		path = os.Args[1]
	}

	ideas := loadIdeas(path)

	sort.Slice(ideas, func(i, j int) bool {
		return ideas[i].Value > ideas[j].Value
	})

	fmt.Println("rank,idea,idea_value")
	for i, idea := range ideas {
		fmt.Printf("%d,%s,%.6f\n", i+1, idea.Name, idea.Value)
	}
}

File: articles/design-thinking-ideation/c/idea_value_engine.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_IDEAS 128
#define MAX_LINE 512
#define MAX_NAME 180

typedef struct {
    char name[MAX_NAME];
    double desirability;
    double feasibility;
    double novelty;
    double equity_value;
    double learning_value;
    double residual_risk;
    double value;
} Idea;

double idea_value(const Idea *idea) {
    const double wd = 0.24;
    const double wf = 0.18;
    const double wn = 0.18;
    const double we = 0.18;
    const double wl = 0.12;
    const double wr = 0.10;

    return wd * idea->desirability
         + wf * idea->feasibility
         + wn * idea->novelty
         + we * idea->equity_value
         + wl * idea->learning_value
         - wr * idea->residual_risk;
}

int compare_desc(const void *a, const void *b) {
    const Idea *ia = (const Idea *)a;
    const Idea *ib = (const Idea *)b;

    if (ia->value < ib->value) return 1;
    if (ia->value > ib->value) return -1;
    return 0;
}

int main(int argc, char **argv) {
    const char *path = argc > 1 ? argv[1] : "../data/raw/idea_portfolio_raw.csv";
    FILE *file = fopen(path, "r");

    if (!file) {
        fprintf(stderr, "ERROR: Unable to open input file: %s\n", path);
        return 1;
    }

    Idea ideas[MAX_IDEAS];
    char line[MAX_LINE];
    int count = 0;

    fgets(line, sizeof(line), file); /* header */

    while (fgets(line, sizeof(line), file) && count < MAX_IDEAS) {
        char *token = strtok(line, ",");
        if (!token) continue;

        strncpy(ideas[count].name, token, MAX_NAME - 1);
        ideas[count].name[MAX_NAME - 1] = '\0';

        strtok(NULL, ","); /* idea_cluster */

        token = strtok(NULL, ",");
        ideas[count].desirability = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        ideas[count].feasibility = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        ideas[count].novelty = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        ideas[count].equity_value = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        ideas[count].learning_value = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        ideas[count].residual_risk = token ? atof(token) : 0.0;

        ideas[count].value = idea_value(&ideas[count]);
        count++;
    }

    fclose(file);

    qsort(ideas, count, sizeof(Idea), compare_desc);

    printf("rank,idea,idea_value\n");
    for (int i = 0; i < count; i++) {
        printf("%d,%s,%.6f\n", i + 1, ideas[i].name, ideas[i].value);
    }

    return 0;
}
